contrib/babelfishpg_tsql/src/pltsql_bulkcopy.c

```diff
@@ -0,0 +1,938 @@
+/*-------------------------------------------------------------------------
+ *
+ * pltsql_bulkcopy.c		- Bulk Copy for PL/tsql
+ *
+ * IDENTIFICATION
+ *	  contrib/babelfishpg_tsql/src/pltsql_bulkcopy.c
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include <ctype.h>
+#include <unistd.h>
+#include <sys/stat.h>
+
+#include "access/heapam.h"
+#include "access/tableam.h"
+#include "access/xact.h"
+#include "catalog/dependency.h"
+#include "catalog/namespace.h"
+#include "commands/sequence.h"
+#include "executor/executor.h"
+#include "executor/nodeModifyTable.h"
+#include "executor/tuptable.h"
+#include "optimizer/optimizer.h"
+#include "miscadmin.h"
+#include "parser/parse_relation.h"
+#include "pltsql_bulkcopy.h"
+#include "rewrite/rewriteHandler.h"
+#include "utils/builtins.h"
+#include "utils/lsyscache.h"
+#include "utils/rel.h"
+#include "utils/rls.h"
+
+/*
+ * No more than this many tuples per CopyMultiInsertBuffer
+ *
+ * Caution: Don't make this too big, as we could end up with this many
+ * CopyMultiInsertBuffer items stored in CopyMultiInsertInfo's
+ * multiInsertBuffers list.
+ */
+#define MAX_BUFFERED_TUPLES		1000
+
+/* Trim the list of buffers back down to this number after flushing */
+#define MAX_PARTITION_BUFFERS	32
+
+/* Stores multi-insert data related to a single relation. */
+typedef struct CopyMultiInsertBuffer
+{
+	TupleTableSlot *slots[MAX_BUFFERED_TUPLES]; /* Array to store tuples */
+	ResultRelInfo *resultRelInfo;	/* ResultRelInfo for 'relid' */
+	BulkInsertState bistate;	/* BulkInsertState for this rel */
+	int			nused;			/* number of 'slots' containing tuples */
+	uint64		linenos[MAX_BUFFERED_TUPLES];	/* Line # of tuple in bulk copy
+												 * stream */
+} CopyMultiInsertBuffer;
+
+/*
+ * Stores one or many CopyMultiInsertBuffers and details about the size and
+ * number of tuples which are stored in them.  This allows multiple buffers to
+ * exist at once when COPYing into a partitioned table.
+ */
+typedef struct CopyMultiInsertInfo
+{
+	List	   *multiInsertBuffers; /* List of tracked CopyMultiInsertBuffers */
+	int			bufferedTuples; /* number of tuples buffered over all buffers */
+	BulkCopyState cstate;		/* Bulk Copy state for this CopyMultiInsertInfo */
+	EState	   *estate;			/* Executor state used for BULK COPY */
+	CommandId	mycid;			/* Command Id used for BULK COPY */
+	int			ti_options;		/* table insert options */
+} CopyMultiInsertInfo;
+
+static BulkCopyState
+BeginBulkCopy(Relation rel,
+			  List *attnamelist);
+
+void BulkCopy(BulkCopyStmt *stmt,
+			uint64 *processed);
+
+void EndBulkCopy(BulkCopyState cstate);
+
+static uint64
+ExecuteBulkCopy(BulkCopyState cstate, int rowCount, int colCount,
+				Datum *Values, bool *Nulls);
+
+static List *
+BulkCopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist);
```

Comment:
extern functions shouldn't be declared again in the c file. You've anyway included the header file.